---
title: "Generalized linear models - Logistic Regression"
author: "Marta Belchior Lopes"
date: "October, 2023"
output:
  html_document:
    number_sections: yes
    toc: yes
  pdf_document:
    number_sections: yes"N"
    toc: yes
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

This practical addresses a binary classification problem using regularized logistic regression to predict the tumor type of lower-grade glioma patients. The data used was extracted form [The Cancer Genome Atlas data portal](https://www.cancer.gov/ccg/research/genome-sequencing/tcga). 

# Matrix construction

## Loading the data

```{r}
# Load a lower-grade glioma (LGG) gene expression (RNA-seq) dataset with 2 class labels: astrocytoma (LGG-a) and oligodendroglioma (LGG-od)

#load("~/LGG_glioma.RData")

## Xdata is the RNA-seq data and Ydata is the response variable with the corresponding class
dim(Xdata)
# RNA-seq data from 381 patients measured over 20501 genes
Xdata[1:5,1:8] # subsampe for matrix visualization
Ydata[1:8]
length(Ydata)
summary(as.factor(Ydata))

# Load R packages
library(ggplot2)
library(glmnet)
library(pROC)
```

## Data preprocessing

```{r}
## Data filtering
# removing samples with standard deviation zero
Xdata_sd <- sapply(seq(ncol(Xdata)), function(ix) {sd(Xdata[,ix])})
Xdata <- Xdata[,Xdata_sd != 0]
dim(Xdata)
# 381 patients measured over 20176 genes

## Data normalization
# computing the z-score
Xdata_sc <- scale(Xdata)
```

# Exploratory data analysis

## Principal component analysis (PCA)

```{r}
# Perform PCA
glioma_pca <- prcomp(Xdata, scale = TRUE)

# Scores of the principal components
scores <- as.data.frame(glioma_pca$x)
scores$class <- Ydata

# Variance explained by each component
variance_explained <- glioma_pca$sdev^2
prop_variance_explained <- variance_explained / sum(variance_explained)
cumulative_prop_variance <- cumsum(prop_variance_explained)

# Plot PCA scores plot
ggplot(scores, aes(x = PC1, y = PC2, color = class)) +
  geom_point() +
  labs(title = paste("PCA Scores plot"),
       x = "PC1",
       y = "PC2") +
theme_minimal()

# Plot % of variance explained
pca_summary <- data.frame(
  PC = 1:length(cumulative_prop_variance),
  Cumulative_Prop_Variance = cumulative_prop_variance
)

ggplot(pca_summary, aes(x = PC, y = Cumulative_Prop_Variance)) +
  geom_line(color = "red") +
  geom_point(color = "gray") +
  labs(title = "PCA: Cumulative Variance Explained",
       x = "Principal Component (PC)",
       y = "Cumulative Proportion of Variance Explained") +
  theme_minimal()

```

# Regularized logistic regression

## Model training

```{r}
# Partition the data into training and test sets
set.seed(2023) # for reproducibility
test_ID <- sample(1:dim(Xdata_sc)[1], round(dim(Xdata_sc)[1]*0.25), replace=FALSE)

Ydata[Ydata=="LGG-a"] <- 1
Ydata[Ydata=="LGG-od"] <- 0

# train set
Xdata_train <- as.matrix(Xdata_sc[-test_ID,])
Ydata_train <- as.factor(Ydata[-test_ID])
# test set
Xdata_test <- as.matrix(Xdata_sc[test_ID,])
Ydata_test <- as.factor(Ydata[test_ID])

# Building the sparse logistic regression model (lambda optimized by cross-validation)

set.seed(1974) # for reproducibility

glioma_fit <- cv.glmnet(Xdata_train, Ydata_train, family="binomial", nfolds=10, alpha=1, type.measure="auc")
glioma_coef <- glioma_fit$glmnet.fit$beta[,which(glioma_fit$cvm == max(glioma_fit$cvm))]
genes_selected <- which(glioma_coef != 0)
length(genes_selected)

data <- data.frame(
  x = 1:length(genes_selected),
  y = glioma_coef[genes_selected],
  labels = names(glioma_coef[genes_selected])
)

ggplot(data, aes(x = x, y = y, label = labels)) +
  geom_point() +
  geom_text(hjust = 0, vjust = 0) +
  labs(title = "Selected variable's coefficients",
       x = "Variable",
       y = "Coefficients") +
  theme_minimal()

## Model predictive performance
# Predicting for the training set

```

## Model evaluation

```{r}
pred_train <- predict(glioma_fit, Xdata_train, type="class", type.measure = "auc", s = "lambda.min")
# Confusion matrix for the train set
table(Ydata_train,pred_train)

# Calculate AUC for the train set
roc_obj <- roc(as.numeric(as.character(Ydata_train)), as.numeric(pred_train))
auc(roc_obj)

# Predicting for a test set
pred_test <- predict(glioma_fit,Xdata_test,type="class")

# Confusion matrix for the test set
table(Ydata_test,pred_test)

# Calculate AUC value for the test set
roc_obj <- roc(as.numeric(as.character(Ydata_test)), as.numeric(pred_test))
auc(roc_obj)

```